File: Lua/color_utils/color_utils_test.lua
--[[
Color Utilities Test Suite

测试颜色工具模块的所有功能。

运行方式: lua color_utils_test.lua
]]

local ColorUtils = require("mod")

-- 测试统计
local testsPassed = 0
local testsFailed = 0
local testsTotal = 0

-- 辅助函数: 断言
local function assertEquals(expected, actual, message)
    testsTotal = testsTotal + 1
    if expected == actual then
        testsPassed = testsPassed + 1
        print("✓ " .. message)
    else
        testsFailed = testsFailed + 1
        print("✗ " .. message .. " (expected: " .. tostring(expected) .. ", got: " .. tostring(actual) .. ")")
    end
end

local function assertApproxEquals(expected, actual, tolerance, message)
    testsTotal = testsTotal + 1
    tolerance = tolerance or 0.01
    if math.abs(expected - actual) <= tolerance then
        testsPassed = testsPassed + 1
        print("✓ " .. message)
    else
        testsFailed = testsFailed + 1
        print("✗ " .. message .. " (expected: " .. tostring(expected) .. "±" .. tolerance .. ", got: " .. tostring(actual) .. ")")
    end
end

local function assertNotNil(value, message)
    testsTotal = testsTotal + 1
    if value ~= nil then
        testsPassed = testsPassed + 1
        print("✓ " .. message)
    else
        testsFailed = testsFailed + 1
        print("✗ " .. message .. " (got nil)")
    end
end

print("=" .. string.rep("=", 59))
print("Color Utilities Test Suite")
print("=" .. string.rep("=", 59))
print("")

-- ============================================================================
-- 测试 RGB 颜色创建
-- ============================================================================
print("--- RGB Color Creation ---")

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    assertEquals(255, color.r, "RGB red component")
    assertEquals(0, color.g, "RGB green component")
    assertEquals(0, color.b, "RGB blue component")
    assertEquals(1, color.a, "RGB default alpha")
end

do
    local color = ColorUtils.fromRgb(128, 64, 32, 0.5)
    assertEquals(128, color.r, "RGB custom red")
    assertEquals(64, color.g, "RGB custom green")
    assertEquals(32, color.b, "RGB custom blue")
    assertEquals(0.5, color.a, "RGB custom alpha")
end

-- ============================================================================
-- 测试 HEX 转换
-- ============================================================================
print("")
print("--- HEX Conversion ---")

do
    local color = ColorUtils.fromHex("#FF5733")
    assertEquals(255, color.r, "HEX red from #FF5733")
    assertEquals(87, color.g, "HEX green from #FF5733")
    assertEquals(51, color.b, "HEX blue from #FF5733")
end

do
    local color = ColorUtils.fromHex("#FFF")
    assertEquals(255, color.r, "HEX short format red")
    assertEquals(255, color.g, "HEX short format green")
    assertEquals(255, color.b, "HEX short format blue")
end

do
    local color = ColorUtils.fromRgb(255, 87, 51)
    assertEquals("#FF5733", color:toHex(), "RGB to HEX conversion")
end

-- ============================================================================
-- 测试 HSL 转换
-- ============================================================================
print("")
print("--- HSL Conversion ---")

do
    local color = ColorUtils.fromHsl(0, 100, 50)  -- 红色
    assertEquals(255, color.r, "HSL red: R")
    assertEquals(0, color.g, "HSL red: G")
    assertEquals(0, color.b, "HSL red: B")
end

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    local hsl = color:toHsl()
    assertEquals(0, hsl.h, "Red to HSL hue")
    assertEquals(100, hsl.s, "Red to HSL saturation")
    assertEquals(50, hsl.l, "Red to HSL lightness")
end

-- ============================================================================
-- 测试 HSV 转换
-- ============================================================================
print("")
print("--- HSV Conversion ---")

do
    local color = ColorUtils.fromHsv(0, 100, 100)  -- 红色
    assertEquals(255, color.r, "HSV red: R")
    assertEquals(0, color.g, "HSV red: G")
    assertEquals(0, color.b, "HSV red: B")
end

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    local hsv = color:toHsv()
    assertEquals(0, hsv.h, "Red to HSV hue")
    assertEquals(100, hsv.s, "Red to HSV saturation")
    assertEquals(100, hsv.v, "Red to HSV value")
end

-- ============================================================================
-- 测试 CMYK 转换
-- ============================================================================
print("")
print("--- CMYK Conversion ---")

do
    local color = ColorUtils.fromCmyk(0, 100, 100, 0)  -- 红色
    assertEquals(255, color.r, "CMYK red: R")
    assertEquals(0, color.g, "CMYK red: G")
    assertEquals(0, color.b, "CMYK red: B")
end

-- ============================================================================
-- 测试命名颜色
-- ============================================================================
print("")
print("--- Named Colors ---")

do
    local color = ColorUtils.fromName("red")
    assertEquals(255, color.r, "Named color 'red': R")
    assertEquals(0, color.g, "Named color 'red': G")
    assertEquals(0, color.b, "Named color 'red': B")
end

-- ============================================================================
-- 测试智能解析
-- ============================================================================
print("")
print("--- Smart Parsing ---")

do
    local color = ColorUtils.parse("#FF0000")
    assertEquals(255, color.r, "Parse HEX")
end

do
    local color = ColorUtils.parse("blue")
    assertEquals(0, color.r, "Parse color name 'blue': R")
    assertEquals(0, color.g, "Parse color name 'blue': G")
    assertEquals(255, color.b, "Parse color name 'blue': B")
end

-- ============================================================================
-- 测试颜色操作
-- ============================================================================
print("")
print("--- Color Operations ---")

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    local gray = color:grayscale()
    assertApproxEquals(gray.r, gray.g, 0, "Grayscale R=G")
    assertApproxEquals(gray.g, gray.b, 0, "Grayscale G=B")
end

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    local inverted = color:invert()
    assertEquals(0, inverted.r, "Invert red: R")
    assertEquals(255, inverted.g, "Invert red: G")
    assertEquals(255, inverted.b, "Invert red: B")
end

-- ============================================================================
-- 测试颜色混合
-- ============================================================================
print("")
print("--- Color Mixing ---")

do
    local red = ColorUtils.fromRgb(255, 0, 0)
    local blue = ColorUtils.fromRgb(0, 0, 255)
    local mixed = red:mix(blue, 0.5)
    assertApproxEquals(127.5, mixed.r, 1, "Mix red and blue: R")
    assertEquals(0, mixed.g, "Mix red and blue: G")
    assertApproxEquals(127.5, mixed.b, 1, "Mix red and blue: B")
end

-- ============================================================================
-- 测试配色方案
-- ============================================================================
print("")
print("--- Color Schemes ---")

do
    local red = ColorUtils.fromRgb(255, 0, 0)
    local comp = red:complementary()
    assertEquals(0, comp.r, "Complementary of red: R")
    assertEquals(255, comp.g, "Complementary of red: G")
    assertEquals(255, comp.b, "Complementary of red: B")
end

do
    local color = ColorUtils.fromRgb(255, 0, 0)
    local triad = color:triadic()
    assertEquals(2, #triad, "Triadic scheme returns 2 colors")
end

-- ============================================================================
-- 测试渐变生成
-- ============================================================================
print("")
print("--- Gradient Generation ---")

do
    local gradient = ColorUtils.gradient({"#FF0000", "#0000FF"}, 5)
    assertEquals(5, #gradient, "Gradient has 5 steps")
    assertEquals(255, gradient[1].r, "Gradient starts with red")
end

-- ============================================================================
-- 测试对比度计算
-- ============================================================================
print("")
print("--- Contrast & WCAG ---")

do
    local white = ColorUtils.fromRgb(255, 255, 255)
    local black = ColorUtils.fromRgb(0, 0, 0)
    local contrast = white:contrastRatio(black)
    assertApproxEquals(21, contrast, 0.1, "Black/white contrast ratio is 21:1")
end

do
    local white = ColorUtils.fromRgb(255, 255, 255)
    local black = ColorUtils.fromRgb(0, 0, 0)
    local compliance = white:wcagCompliance(black)
    assertEquals(true, compliance.aaNormal, "Black/white passes AA Normal")
end

-- ============================================================================
-- 测试随机颜色
-- ============================================================================
print("")
print("--- Random Colors ---")

do
    local color = ColorUtils.random()
    assertNotNil(color.r, "Random color has R")
    assertNotNil(color.g, "Random color has G")
    assertNotNil(color.b, "Random color has B")
end

do
    local color1 = ColorUtils.fromSeed(42)
    local color2 = ColorUtils.fromSeed(42)
    assertEquals(color1.r, color2.r, "Seeded colors have same R")
    assertEquals(color1.g, color2.g, "Seeded colors have same G")
end

-- ============================================================================
-- 测试实用工具
-- ============================================================================
print("")
print("--- Utility Functions ---")

do
    local name = ColorUtils.findClosestName(ColorUtils.fromRgb(255, 0, 0))
    assertEquals("red", name, "Find closest name for red")
end

do
    local isLight = ColorUtils.isLight("#FFFFFF")
    assertEquals(true, isLight, "White is light")
end

do
    local contrast = ColorUtils.getContrastColor("#000000")
    assertEquals(255, contrast.r, "Contrast for black is white: R")
end

do
    local temp = ColorUtils.getTemperature("#FF0000")
    assertEquals("warm", temp, "Red is warm")
end

-- ============================================================================
-- 测试边界情况
-- ============================================================================
print("")
print("--- Edge Cases ---")

do
    local color = ColorUtils.fromRgb(300, -50, 128)
    assertEquals(255, color.r, "Clamp R to 255")
    assertEquals(0, color.g, "Clamp G to 0")
end

-- ============================================================================
-- 测试结果汇总
-- ============================================================================
print("")
print("=" .. string.rep("=", 59))
print(string.format("Test Results: %d passed, %d failed, %d total", testsPassed, testsFailed, testsTotal))
print("=" .. string.rep("=", 59))

if testsFailed > 0 then
    os.exit(1)
else
    print("All tests passed! ✓")
    os.exit(0)
end

File: Swift/color_utils/ColorUtilsTest.swift
/**
 * AllToolkit - Swift Color Utilities Tests
 *
 * 测试颜色工具类的各种功能
 */

import Foundation

// 简单测试框架
class TestRunner {
    private var passed = 0
    private var failed = 0
    private var tests: [(name: String, test: () -> Bool)] = []
    
    func addTest(_ name: String, _ test: @escaping () -> Bool) {
        tests.append((name, test))
    }
    
    func run() {
        print("=" + String(repeating: "=", count: 59) + "=")
        print("Swift Color Utilities Tests")
        print("=" + String(repeating: "=", count: 59) + "=")
        print()
        
        for (name, test) in tests {
            let result = test()
            let status = result ? "✅ PASS" : "❌ FAIL"
            print("\(status) - \(name)")
            if result {
                passed += 1
            } else {
                failed += 1
            }
        }
        
        print()
        print("=" + String(repeating: "=", count: 59) + "=")
        print("Results: \(passed) passed, \(failed) failed")
        print("=" + String(repeating: "=", count: 59) + "=")
    }
}

func assertEquals<T: Equatable>(_ expected: T, _ actual: T, _ message: String = "") -> Bool {
    if expected == actual {
        return true
    } else {
        if !message.isEmpty {
            print("  Expected: \(expected), Got: \(actual) - \(message)")
        } else {
            print("  Expected: \(expected), Got: \(actual)")
        }
        return false
    }
}

func assertApproxEquals(_ expected: Double, _ actual: Double, _ tolerance: Double = 0.01, _ message: String = "") -> Bool {
    if abs(expected - actual) <= tolerance {
        return true
    } else {
        if !message.isEmpty {
            print("  Expected: \(expected), Got: \(actual) (tolerance: \(tolerance)) - \(message)")
        } else {
            print("  Expected: \(expected), Got: \(actual) (tolerance: \(tolerance))")
        }
        return false
    }
}

func assertTrue(_ condition: Bool, _ message: String = "") -> Bool {
    if condition {
        return true
    } else {
        print("  Assertion failed: \(message)")
        return false
    }
}

// 创建测试
let runner = TestRunner()

// MARK: - Color Creation Tests

runner.addTest("Color: Basic RGB creation") {
    let color = Color(red: 1, green: 0.5, blue: 0, alpha: 1)
    return assertEquals(1.0, color.red) &&
           assertEquals(0.5, color.green) &&
           assertEquals(0.0, color.blue) &&
           assertEquals(1.0, color.alpha)
}

runner.addTest("Color: RGB Int creation") {
    let color = Color(r: 255, g: 128, b: 0)
    return assertApproxEquals(1.0, color.red) &&
           assertApproxEquals(128.0/255.0, color.green) &&
           assertApproxEquals(0.0, color.blue)
}

runner.addTest("Color: Predefined colors") {
    return assertEquals(Color.black, Color(red: 0, green: 0, blue: 0)) &&
           assertEquals(Color.white, Color(red: 1, green: 1, blue: 1)) &&
           assertEquals(Color.red, Color(red: 1, green: 0, blue: 0))
}

// MARK: - HEX Conversion Tests

runner.addTest("HEX: Create from 6-char hex string") {
    let color = Color(hex: "#FF0000")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red) &&
           assertApproxEquals(0.0, color!.green) &&
           assertApproxEquals(0.0, color!.blue)
}

runner.addTest("HEX: Create from 3-char hex string") {
    let color = Color(hex: "#F00")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red) &&
           assertApproxEquals(0.0, color!.green) &&
           assertApproxEquals(0.0, color!.blue)
}

runner.addTest("HEX: Create from hex integer") {
    let color = Color(hex: 0x00FF00)
    return assertApproxEquals(0.0, color.red) &&
           assertApproxEquals(1.0, color.green) &&
           assertApproxEquals(0.0, color.blue)
}

runner.addTest("HEX: Convert to hex string") {
    let color = Color(red: 1, green: 0.5, blue: 0)
    return assertEquals("#FF8000", color.toHex())
}

runner.addTest("HEX: Convert to hex int") {
    let color = Color(red: 1, green: 0, blue: 0)
    return assertEquals(UInt32(0xFF0000), color.toHexInt())
}

// MARK: - HSL Conversion Tests

runner.addTest("HSL: Convert to HSL") {
    let color = Color(red: 1, green: 0, blue: 0)
    let hsl = color.hsl
    return assertApproxEquals(0, hsl.hue) &&
           assertApproxEquals(1.0, hsl.saturation) &&
           assertApproxEquals(0.5, hsl.lightness)
}

runner.addTest("HSL: Create from HSL") {
    let color = Color(hue: 120, saturation: 1.0, lightness: 0.5)
    return assertApproxEquals(0.0, color.red, 0.02) &&
           assertApproxEquals(1.0, color.green, 0.02) &&
           assertApproxEquals(0.0, color.blue, 0.02)
}

runner.addTest("HSL: Grayscale conversion") {
    let color = Color(red: 1, green: 0, blue: 0)
    let gray = color.grayscale()
    return assertApproxEquals(gray.red, gray.green) &&
           assertApproxEquals(gray.green, gray.blue)
}

// MARK: - HSV Conversion Tests

runner.addTest("HSV: Convert to HSV") {
    let color = Color(red: 1, green: 0, blue: 0)
    let hsv = color.hsv
    return assertApproxEquals(0, hsv.hue) &&
           assertApproxEquals(1.0, hsv.saturation) &&
           assertApproxEquals(1.0, hsv.value)
}

runner.addTest("HSV: Create from HSV") {
    let color = Color(hue: 240, saturation: 1.0, value: 1.0)
    return assertApproxEquals(0.0, color.red, 0.02) &&
           assertApproxEquals(0.0, color.green, 0.02) &&
           assertApproxEquals(1.0, color.blue, 0.02)
}

// MARK: - CMYK Conversion Tests

runner.addTest("CMYK: Convert to CMYK") {
    let color = Color(red: 1, green: 0, blue: 0)
    let cmyk = color.cmyk
    return assertApproxEquals(0.0, cmyk.cyan) &&
           assertApproxEquals(1.0, cmyk.magenta) &&
           assertApproxEquals(1.0, cmyk.yellow) &&
           assertApproxEquals(0.0, cmyk.key)
}

runner.addTest("CMYK: Create from CMYK") {
    let color = Color(cyan: 0, magenta: 1, yellow: 1, key: 0)
    return assertApproxEquals(1.0, color.red, 0.02) &&
           assertApproxEquals(0.0, color.green, 0.02) &&
           assertApproxEquals(0.0, color.blue, 0.02)
}

// MARK: - Color Manipulation Tests

runner.addTest("Manipulation: Lighten color") {
    let color = Color(red: 0.5, green: 0.5, blue: 0.5)
    let lightened = color.lighten(0.2)
    let hsl = lightened.hsl
    return assertApproxEquals(0.7, hsl.lightness)
}

runner.addTest("Manipulation: Darken color") {
    let color = Color(red: 0.5, green: 0.5, blue: 0.5)
    let darkened = color.darken(0.2)
    let hsl = darkened.hsl
    return assertApproxEquals(0.3, hsl.lightness)
}

runner.addTest("Manipulation: Invert color") {
    let color = Color(red: 0.2, green: 0.4, blue: 0.6)
    let inverted = color.inverted()
    return assertApproxEquals(0.8, inverted.red) &&
           assertApproxEquals(0.6, inverted.green) &&
           assertApproxEquals(0.4, inverted.blue)
}

runner.addTest("Manipulation: Saturate color") {
    let color = Color(hue: 0, saturation: 0.5, lightness: 0.5)
    let saturated = color.saturate(0.3)
    return assertApproxEquals(0.8, saturated.hsl.saturation)
}

runner.addTest("Manipulation: Desaturate color") {
    let color = Color(hue: 0, saturation: 0.5, lightness: 0.5)
    let desaturated = color.desaturate(0.3)
    return assertApproxEquals(0.2, desaturated.hsl.saturation)
}

// MARK: - Color Mixing Tests

runner.addTest("Mixing: Mix two colors equally") {
    let color1 = Color(red: 1, green: 0, blue: 0)
    let color2 = Color(red: 0, green: 0, blue: 1)
    let mixed = color1.mix(with: color2, ratio: 0.5)
    return assertApproxEquals(0.5, mixed.red) &&
           assertApproxEquals(0.0, mixed.green) &&
           assertApproxEquals(0.5, mixed.blue)
}

runner.addTest("Mixing: Tint (mix with white)") {
    let color = Color(red: 1, green: 0, blue: 0)
    let tinted = color.tint(0.5)
    return assertApproxEquals(1.0, tinted.red) &&
           assertApproxEquals(0.5, tinted.green) &&
           assertApproxEquals(0.5, tinted.blue)
}

runner.addTest("Mixing: Shade (mix with black)") {
    let color = Color(red: 1, green: 0, blue: 0)
    let shaded = color.shade(0.5)
    return assertApproxEquals(0.5, shaded.red) &&
           assertApproxEquals(0.0, shaded.green) &&
           assertApproxEquals(0.0, shaded.blue)
}

runner.addTest("Mixing: Weighted mix") {
    let colors: [(color: Color, weight: Double)] = [
        (Color(red: 1, green: 0, blue: 0), 1.0),
        (Color(red: 0, green: 1, blue: 0), 1.0),
        (Color(red: 0, green: 0, blue: 1), 1.0)
    ]
    let avg = Color.weightedMix(colors)
    return assertTrue(avg != nil) &&
           assertApproxEquals(1.0/3.0, avg!.red, 0.02) &&
           assertApproxEquals(1.0/3.0, avg!.green, 0.02) &&
           assertApproxEquals(1.0/3.0, avg!.blue, 0.02)
}

// MARK: - Contrast Tests

runner.addTest("Contrast: Relative luminance of white") {
    let white = Color.white
    return assertApproxEquals(1.0, white.relativeLuminance)
}

runner.addTest("Contrast: Relative luminance of black") {
    let black = Color.black
    return assertApproxEquals(0.0, black.relativeLuminance)
}

runner.addTest("Contrast: Contrast ratio black on white") {
    let ratio = Color.black.contrastRatio(with: .white)
    return assertApproxEquals(21.0, ratio, 0.1)
}

runner.addTest("Contrast: Is light/dark detection") {
    return assertTrue(Color.white.isLight) &&
           assertTrue(Color.black.isDark)
}

runner.addTest("Contrast: Text color selection") {
    let darkBg = Color.black
    let lightBg = Color.white
    return assertEquals(Color.white, darkBg.textColor) &&
           assertEquals(Color.black, lightBg.textColor)
}

runner.addTest("Contrast: WCAG AA pass check") {
    let ratio = Color.black.contrastRatio(with: .white)
    return assertTrue(ratio >= 4.5, "Black on white should pass WCAG AA")
}

// MARK: - Palette Generation Tests

runner.addTest("Palette: Analogous palette") {
    let base = Color(red: 1, green: 0, blue: 0)
    let palette = base.analogousPalette(count: 5)
    return assertEquals(5, palette.count) &&
           assertTrue(palette.allSatisfy { $0.alpha == 1.0 })
}

runner.addTest("Palette: Complementary palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = base.complementaryPalette()
    return assertEquals(2, palette.count) &&
           assertApproxEquals(180, palette[1].hsl.hue, 1.0)
}

runner.addTest("Palette: Triadic palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = base.triadicPalette()
    return assertEquals(3, palette.count)
}

runner.addTest("Palette: Split complementary palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = base.splitComplementaryPalette()
    return assertEquals(3, palette.count)
}

runner.addTest("Palette: Tetradic palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = base.tetradicPalette()
    return assertEquals(4, palette.count)
}

runner.addTest("Palette: Monochromatic palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = base.monochromaticPalette(steps: 5)
    return assertEquals(5, palette.count)
}

// MARK: - Random Color Tests

runner.addTest("Random: Generate random color") {
    let color1 = Color.random()
    let color2 = Color.random()
    return assertTrue(color1.alpha == 1.0) &&
           assertTrue(color2.alpha == 1.0) &&
           assertTrue(color1 != color2 || true) // Colors might be equal by chance
}

runner.addTest("Random: Generate random color with alpha") {
    let color = Color.random(opaque: false)
    return assertTrue(color.alpha >= 0 && color.alpha <= 1)
}

runner.addTest("Random: Seed-based color generation") {
    let color1 = Color.fromSeed("test")
    let color2 = Color.fromSeed("test")
    let color3 = Color.fromSeed("different")
    return assertEquals(color1, color2) &&
           assertTrue(color1 != color3 || true)
}

// MARK: - CSS Name Tests

runner.addTest("CSS: Create from CSS name") {
    let red = Color(cssName: "red")
    let blue = Color(cssName: "blue")
    return assertTrue(red != nil && blue != nil) &&
           assertApproxEquals(1.0, red!.red) &&
           assertApproxEquals(1.0, blue!.blue)
}

runner.addTest("CSS: Convert to CSS name") {
    let red = Color(red: 1, green: 0, blue: 0)
    let white = Color.white
    return assertEquals("red", red.cssName) &&
           assertEquals("white", white.cssName)
}

runner.addTest("CSS: Case insensitive name parsing") {
    let red1 = Color(cssName: "RED")
    let red2 = Color(cssName: "Red")
    return assertTrue(red1 != nil && red2 != nil) &&
           assertEquals(red1, red2)
}

// MARK: - ColorUtils Tests

runner.addTest("Utils: Parse hex string") {
    let color = ColorUtils.parse("#FF0000")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red) &&
           assertApproxEquals(0.0, color!.green) &&
           assertApproxEquals(0.0, color!.blue)
}

runner.addTest("Utils: Parse rgb string") {
    let color = ColorUtils.parse("rgb(255, 0, 0)")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red) &&
           assertApproxEquals(0.0, color!.green) &&
           assertApproxEquals(0.0, color!.blue)
}

runner.addTest("Utils: Parse rgba string") {
    let color = ColorUtils.parse("rgba(255, 0, 0, 0.5)")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red) &&
           assertApproxEquals(0.5, color!.alpha)
}

runner.addTest("Utils: Parse hsl string") {
    let color = ColorUtils.parse("hsl(120, 100%, 50%)")
    return assertTrue(color != nil) &&
           assertApproxEquals(0.0, color!.red, 0.02) &&
           assertApproxEquals(1.0, color!.green, 0.02)
}

runner.addTest("Utils: Parse CSS name") {
    let color = ColorUtils.parse("red")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red)
}

runner.addTest("Utils: Gradient generation") {
    let gradient = ColorUtils.gradient(from: .black, to: .white, steps: 5)
    return assertEquals(5, gradient.count) &&
           assertEquals(Color.black, gradient.first) &&
           assertEquals(Color.white, gradient.last)
}

runner.addTest("Utils: Material palette") {
    let base = Color(red: 0.2, green: 0.4, blue: 0.8)
    let palette = ColorUtils.materialPalette(from: base)
    return assertEquals(10, palette.count) &&
           assertTrue(palette["500"] != nil)
}

runner.addTest("Utils: Best contrast selection") {
    let bg = Color.white
    let candidates = [Color.black, Color.gray, Color.darkGray]
    let best = ColorUtils.bestContrast(for: bg, among: candidates)
    return assertEquals(Color.black, best)
}

runner.addTest("Utils: Random palette") {
    let palette = ColorUtils.randomPalette(count: 5)
    return assertEquals(5, palette.count)
}

runner.addTest("Utils: Harmonious palette") {
    let base = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let palette = ColorUtils.harmoniousPalette(from: base)
    return assertEquals(5, palette.count)
}

// MARK: - RGB/RGBA String Tests

runner.addTest("String: RGB string format") {
    let color = Color(red: 1, green: 0.5, blue: 0)
    return assertEquals("rgb(255, 128, 0)", color.rgbString)
}

runner.addTest("String: RGBA string format") {
    let color = Color(red: 1, green: 0.5, blue: 0, alpha: 0.5)
    return assertTrue(color.rgbaString.contains("rgba(255, 128, 0"))
}

// MARK: - Edge Cases Tests

runner.addTest("Edge: Hex without # prefix") {
    let color = Color(hex: "FF0000")
    return assertTrue(color != nil) &&
           assertApproxEquals(1.0, color!.red)
}

runner.addTest("Edge: Invalid hex returns nil") {
    let color = Color(hex: "invalid")
    return assertTrue(color == nil)
}

runner.addTest("Edge: RGB clamping - values above 255") {
    let color = Color(r: 300, g: 128, b: 0)
    return assertApproxEquals(1.0, color.red) &&
           assertApproxEquals(128.0/255.0, color.green)
}

runner.addTest("Edge: RGB clamping - negative values") {
    let color = Color(r: -50, g: 0, b: 0)
    return assertApproxEquals(0.0, color.red)
}

runner.addTest("Edge: Alpha clamping") {
    let color = Color(red: 0.5, green: 0.5, blue: 0.5, alpha: 2.0)
    return assertApproxEquals(1.0, color.alpha)
}

runner.addTest("Edge: HSL hue wrapping") {
    let color1 = Color(hue: 0, saturation: 1.0, lightness: 0.5)
    let color2 = Color(hue: 360, saturation: 1.0, lightness: 0.5)
    let color3 = Color(hue: 720, saturation: 1.0, lightness: 0.5)
    return assertApproxEquals(color1.red, color2.red, 0.02) &&
           assertApproxEquals(color1.red, color3.red, 0.02)
}

// 运行测试
runner.run()